src/tree_data.c

```diff
@@ -2847,7 +2938,16 @@ lyd_compare_single(const struct lyd_node *node1, const struct lyd_node *node2, u
 
             term1 = (struct lyd_node_term *)node1;
             term2 = (struct lyd_node_term *)node2;
-            return term1->value.realtype->plugin->compare(&term1->value, &term2->value);
+
+            if (LYD_CTX(node1) == LYD_CTX(node2)) {
+                return term1->value.realtype->plugin->compare(&term1->value, &term2->value);
+            } else {
+                if (strcmp(LYD_CANON_VALUE(node1), LYD_CANON_VALUE(node2))) {
+                    return LY_ENOT;
+                } else {
+                    return LY_SUCCESS;
+                }
+            }
```

Comment:
It took me a few seconds to notice that `break` is not missing because a `return` must be called but you can save these seconds for any next person by refactoring this code a bit. Something like this would be sufficient but it is up to you to improve it.
```C
            if (LYD_CTX(node1) == LYD_CTX(node2)) {
                return term1->value.realtype->plugin->compare(&term1->value, &term2->value);
            }

            /* different context */
            if (strcmp(LYD_CANON_VALUE(node1), LYD_CANON_VALUE(node2))) {
                return LY_ENOT;
            } else {
                return LY_SUCCESS;
            }
```
One `if-else` is fine, but not nested.

Reply:
ok